### PR TITLE
Api

### PR DESCRIPTION
**Pull request by ilyaZamkovoy**

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -8,6 +8,9 @@ gem "rename", "~> 1.0", ">= 1.0.2"
 gem "sprockets", "3.6.3"
 gem "ruby-beautify", "~> 0.97.3"
 gem "html_truncator", "~>0.2"
+gem "nokogiri", "1.6.6.1"
+gem "simple_token_authentication"
+gem "devise-token_authenticatable"
```
Comment (vladimirbazhanov) [inline Gemfile]: Are you sure you need both gems for token auth? May need really need only one? )

File: app/controllers/api/v1/registrations_controller.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
+class Api::V1::RegistrationsController < Devise::RegistrationsController
+  respond_to :json
+
+  def create
+    user = User.new(user_params)
+    if user.save
+    	sign_in(user)
+      render :json=> user.as_json(:auth_token=>user.auth_token, :email=>user.email), :status=>201
```
Comment (vladimirbazhanov) [inline app/controllers/api/v1/registrations_controller.rb]: You goal is to make this work
```
render user
```
or
```
respond_with user
```

File: app/controllers/api/v1/registrations_controller.rb [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+      return
+    else
+      warden.custom_failure!
+      render :json=> user.errors, :status=>422
```
Comment (vladimirbazhanov) [inline app/controllers/api/v1/registrations_controller.rb]: render user.errors should be here :)

File: app/controllers/api/v1/sessions_controller.rb [hunk 4]
```diff
@@ -0,0 +1,43 @@
+class Api::V1::SessionsController < Devise::SessionsController
+  prepend_before_filter :require_no_authentication, :only => [:create ]
```
Comment (vladimirbazhanov) [inline app/controllers/api/v1/sessions_controller.rb]: pls dont you old syntax, use
```
only: %i(create)
```

File: app/models/user.rb [hunk 5]
```diff
@@ -10,12 +10,11 @@ class User < ActiveRecord::Base
...
 
-  after_create :add_token
+  before_create :add_token
 
   private
 
   def add_token
-    return if token.present?
-    self.token = SecureRandom.hex
+    self.auth_token = SecureRandom.hex
```
Comment (vladimirbazhanov) [inline app/models/user.rb]: you dont need self here

File: db/migrate/20161128114741_add_index_to_users_token.rb [hunk 6]
```diff
@@ -0,0 +1,5 @@
+class AddIndexToUsersToken < ActiveRecord::Migration
+  def change
+  	add_index  :users, :auth_token, unique: true
```
Comment (vladimirbazhanov) [inline db/migrate/20161128114741_add_index_to_users_token.rb]: 2 spaces

File: db/schema.rb [hunk 7]
```diff
@@ -64,9 +64,10 @@
     t.datetime "created_at"
     t.datetime "updated_at"
     t.string   "full_name"
-    t.string   "token"
+    t.string   "auth_token"
   end
 
+  add_index "users", ["auth_token"], name: "index_users_on_auth_token", unique: true, using: :btree
   add_index "users", ["email"], name: "index_users_on_email", unique: true, using: :btree
   add_index "users", ["reset_password_token"], name: "index_users_on_reset_password_token", unique: true, using: :btree
```
Comment (vladimirbazhanov) [inline db/schema.rb]: also pls remove file 'server.key' from repository

File: app/controllers/api/v1/sessions_controller.rb [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+      sign_in(@current_user)
+      render json: {success: true, auth_token: @current_user.auth_token, email: @current_user.email}
+      return
+    end
+    invalid_login_attempt
+  end
+
+  def destroy
+    sign_out(@current_user)
+    render json: {message: "sign_out susccessfull"}
```
Comment (vladimirbazhanov) [inline app/controllers/api/v1/sessions_controller.rb]: I think you really dont' need this message

If 'sign out' request was successfull you can return only http code, for example 204 that means “No Content” success code, for DELETE request.

File: app/controllers/api/v1/sessions_controller.rb [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+  end
+
+  protected
+  def ensure_params_exist
+    return unless params[:email].blank?
+    render :json=>{:success=>false, :message=>"missing user_email parameter"}, status: 422
```
Comment (vladimirbazhanov) [inline app/controllers/api/v1/sessions_controller.rb]: Why you need this check? I'm sure you really need only try to sign in user and return 404 if user was not found. All other check can be removed

Why you checking for email presence? You iOS mobile app must check email presence itself and dont request to server if user did not enter email

File: Gemfile (hunk 1 above)
Comment (ilyaZamkovoy) [inline Gemfile]: nokogiri really not using

File: app/controllers/api/v1/posts_controller.rb [hunk 10]
```diff
@@ -1,6 +1,33 @@
-class Api::V1::PostsController < ApplicationController
+class Api::V1::PostsController < Api::V1::ApplicationController
   def show
+    @post = Post.search(params.fetch(:qs, ""))
```
Comment (vladimirbazhanov) [inline app/controllers/api/v1/posts_controller.rb]: route for this action is GET /posts/:id/ so you have ID of post and you need to
```
@post = Post.find(params[:id])
```

File: db/schema.rb (hunk 7 above)
Comment (vladimirbazhanov) [inline db/schema.rb]: file 'server.key' still here :)

File: app/controllers/api/v1/comments_controller.rb [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+      @comment.update_attributes(comment_params)
+      respond_with @comment
+    else
+      render json: { message: "its not your comment" }.to_json
+    end
+  end
+
+  def destroy
+    @comment = Comment.find(params[:id])
+    if policy(@comment).destroy?
```
Comment (vladimirbazhanov) [inline app/controllers/api/v1/comments_controller.rb]: try to use ```authorize``` method like in examples

```
def update
  @post = Post.find(params[:id])
  authorize @post
  if @post.update(post_params)
    redirect_to @post
  else
    render :edit
  end
end
```

Comment (ilyaZamkovoy) [inline app/controllers/api/v1/comments_controller.rb]: def update
    @comment = Comment.find(params[:id])
    authorize @comment
    if @comment.update?
      @comment.update_attributes(comment_params)
      respond_with @comment
    else
      render json: { message: "its not your comment" }.to_json
    end
  end


is that correct?

File: app/controllers/api/v1/comments_controller.rb [hunk 12]
```diff
@@ -1,4 +1,5 @@
 class Api::V1::CommentsController < Api::V1::ApplicationController
+  after_action :verify_authorized
```
Comment (vladimirbazhanov) [inline app/controllers/api/v1/comments_controller.rb]: remove this

File: app/controllers/api/v1/concerns/authorization.rb [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+
+  included do
+    include Pundit
+    rescue_from Pundit::NotAuthorizedError, with: :user_not_authorized
+  end
+
+  private
+
+  def user_not_authorized
+    render json: { message: "its not your comment" }.to_json
```
Comment (vladimirbazhanov) [inline app/controllers/api/v1/concerns/authorization.rb]: its not your comment -> you have not authorized for this action

401 Unathorized

File: app/controllers/api/v1/posts_controller.rb [hunk 14]
```diff
@@ -25,11 +25,11 @@ def update
 
   def destroy
     @post = Post.find(params[:id])
-    if policy(@post).update?
-      @post.update_attributes(post_params)
-      respond_with @post
-    else
+    authorize @post
+    if @post.destroy
       render json: { message: "post successfully deleted" }.to_json
```
Comment (vladimirbazhanov) [inline app/controllers/api/v1/posts_controller.rb]: do not use messages

**Pull request merged**